src/lsm/fuzz.zig

```diff
@@ -0,0 +1,459 @@
+        assertion.verify();
+        while (assertion.verify_count > 0) try env.tick();
+    }
+
+    fn run(storage: *Storage, fuzz_ops: []const FuzzOp) !void {
+        var env: Environment = undefined;
+
+        try env.init(storage);
+        defer env.deinit();
+
+        // Open the superblock then forest.
+        try env.open();
+
+        // The forest should behave like a simple key-value data-structure.
+        // We'll compare it to a hash map.
+        var model = std.hash_map.AutoHashMap(u128, Account).init(allocator);
+        defer model.deinit();
+
+        // We need an op number for compact.
+        var op: u64 = 1;
+
+        for (fuzz_ops) |fuzz_op, fuzz_op_index| {
+            log.debug("Running fuzz_ops[{}] == {}", .{ fuzz_op_index, fuzz_op });
+            // Apply fuzz_op to the forest and the model.
+            switch (fuzz_op) {
+                .compact => {
+                    try env.compact(op);
+                    op += 1;
+                },
+                .put => |account| {
+                    model.put(account.id, account) catch unreachable;
+                    const groove = &env.forest.grooves.accounts;
+                    groove.put(&account);
+                },
+                .get => |get| {
+                    const model_account = model.get(get.id);
+                    const groove = &env.forest.grooves.accounts;
+                    const lsm_account = groove.get(get.id);
+                    if (model_account == null)
+                        assert(lsm_account == null)
+                    else
+                        assert(model_account.?.equal(lsm_account.?));
+                },
+            }
+        }
+    }
+};
+
+fn randomIntExp(comptime Int: type, random: std.rand.Random, avg_int: Int) Int {
+    return @floatToInt(Int, @trunc(random.floatExp(f64) * @intToFloat(f64, avg_int)));
+}
+
+pub fn fuzz(fuzz_ops: []const FuzzOp) !void {
+    // Init mocked storage.
+    var storage = try Storage.init(
+        allocator,
+        Environment.size_max +
+            // TODO this is totally arbitrary
+            16 * 1024 * 1024 * 1024,
+        Storage.Options{
+            // We don't apply storage faults yet, so this seed doesn't matter.
+            .seed = 0xdeadbeef,
+            .read_latency_min = 0,
+            .read_latency_mean = 0,
+            .write_latency_min = 0,
+            .write_latency_mean = 0,
+            .read_fault_probability = 0,
+            .write_fault_probability = 0,
+        },
+        0,
+        .{
+            .first_offset = 0,
+            .period = 0,
+        },
+    );
+    defer storage.deinit(allocator);
+
+    try Environment.format(&storage);
+    try Environment.run(&storage, fuzz_ops);
+}
+
+pub fn generate_fuzz_ops(random: std.rand.Random) ![]const FuzzOp {
+    const fuzz_op_count = randomIntExp(usize, random, 1E6);
+    const fuzz_ops = try allocator.alloc(FuzzOp, fuzz_op_count);
```

Comment:
Even when there are no other `try`'s that could trigger it, we always include a matching `errdefer`. That way if the function is modified, it is already correct & we can't forget it by mistake.

```suggestion
    const fuzz_ops = try allocator.alloc(FuzzOp, fuzz_op_count);
    errdefer allocator.free(fuzz_ops);
```